### ISSUE TITLE
Migration of modular file system plugins in tensorflow (s3/hdfs/gcs) to tensorflow-io

### ISSUE THREAD

**Issue by yongtang** [tensorflow/io#1183]
As was discussed in our September monthly meetings, we are looking into migrate modular file system plugins in tensorflow (s3/hdfs/gcs) to tensorflow-io package.

Once the migration is done the size and build time of tensorflow repo could be substantially improved, this can greatly help external tensorflow contributors experience as they could selectively decide to only build some components (e.g., tensorflow core, tensorflow-io, tensorflow-addons, etc).

Currently it takes at least 8 hours (or even more) for many external tensorflow contributors to build complete tensorflow package due to the lack of bazel cache that is only internally available in google.

Now as we already have a couple of modular file systems in tensorflow-io (azure blob file system and http file system. see #1111), it might be time to start looking into migration of plugins in tensorflow to tensorflow-io.

/cc @mihaimaruseac @vnvo2409

**Comment by vnvo2409** [comment]
Let me check if there is any incompatible between the core implementation and the plugins.

There are some points which I would like to make it clear.

---

Documentation: There are some issues of TF related to the behavior of the environment variables: 
- `S3`:  https://github.com/tensorflow/tensorflow/issues/44564 https://github.com/tensorflow/tensorflow/issues/42389 
- `GCS`: https://github.com/tensorflow/tensorflow/issues/15530

Up till now, I don't see much documentation about those filesystems. I would like to have a working example and documentation about other envs. Where should I make a PR to ?

---

Issues: Could we transfer the issues related to those filesystems from TF repo to this repo ?

---

Testing: What we should do about testing ? I think we could set up a `s3` and `gcs` emulator in `github actions`. Another possibility is tagging those tests `manual` and run it locally.

**Comment by yongtang** [comment]
Thanks @vnvo2409 !

For testing: I have created a PR #1184 which setup the AWS Localstack emulator in GitHub Actions. See https://github.com/tensorflow/io/blob/master/tests/test_aws/aws_test.sh for localstack emulator setup.

The test is working for Kinesis dataset in tensorflow-io. I configured the localstack emulator to have s3 and a bare-bone stub test has been created as well as a starting. Please take a look (https://github.com/tensorflow/io/blob/master/tests/test_s3_eager.py)

I will take a look at gcs emulator and see if a similar setup could be done as well.

For issue transfer: Will take a look to see if issue can be transferred here.

For documentation: At the moment the documentation source is inside the `docs` directory of this repo (https://github.com/tensorflow/io/tree/master/docs). The PRs are reviewed by docs team. PR #1172 is an example.

Once PR is merged the content are being pushed to tensorflow.org website (see https://www.tensorflow.org/io/overview for example).

**Comment by vnvo2409** [comment]
> For testing: I have created a PR #1184 which setup the AWS Localstack emulator in GitHub Actions.

Thank you!

> I will take a look at gcs emulator and see if a similar setup could be done as well.

I think we could use the [emulator](https://github.com/googleapis/google-cloud-cpp/tree/master/google/cloud/storage/emulator) of `google-cloud-cpp`, which is the library underlies gcs plugins. The emulator isn't done yet ( It will be done around mid-november ). It is written entirely in python so setup should be quite easy.

I am a little busy working with the emulator of `google-cloud-cpp`, I will take over this migration when I finish the other work.

**Comment by yongtang** [comment]
@vnvo2409 We don't have a hard timeline for it though we are looking at before TF 2.5 release. Since 2.4 is not released yet (likely in several weeks), the TF 2.5 release will be about early next year. I will try to take a look and see if I can help migrating a plugin or so, and may ask you for review 😄 .

**Comment by vnvo2409** [comment]
> the TF 2.5 release will be about early next year

Early next year should be enough for me.

> I will try to take a look and see if I can help migrating a plugin or so, and may ask you for review 😄 .

Thank you very much 😄

**Comment by mihaimaruseac** [comment]
I could help with moving issues.

Thank you for making the tracking issue to move the plugins.

**Comment by yongtang** [comment]
@mihaimaruseac One issue I noticed is that at the moment there is no way to `unload` the existing file systems embedded in TF. For example, if I am trying to register a shared object plugin for S3, the following error will thrown out because there is already a `s3` registered:
```
/Library/Python/3.8/site-packages/tensorflow/python/framework/load_library.py:178: in register_filesystem_plugin
    py_tf.TF_RegisterFilesystemPlugin(plugin_location)
E   tensorflow.python.framework.errors_impl.AlreadyExistsError: File system for s3 already registered
```

It looks like the error arises from the code in https://github.com/tensorflow/tensorflow/blob/dc3099c444d294b39cd79fe1d1a4bff59a0c6180/tensorflow/core/platform/env.cc#L82-L90

I am wondering if it makes sense to allow plugin registration to override existing static built file systems (s3/gcs/hdf5)? I think that will allow user to have more flexibilities. (We can lock `local`, `unix` file system to never be overridden).

**Comment by mihaimaruseac** [comment]
We should comment the `TF_REGISTER_FILESYSTEM` lines in the old code for every filesystem that is integrally moved to plugins. In fact, we can also delete the non-modular implementation.

It will be a risky thing to allow overriding filesystem registrations.

**Comment by kvignesh1420** [comment]
@vnvo2409 @mihaimaruseac what is the current approach for testing `hdfs` plugins in tensorflow?

**Comment by vnvo2409** [comment]
AFAIK, we could run `hdfs` locally in pseudo-distributed mode and it should be enough for testing

**Comment by kvignesh1420** [comment]
how about `hdfs` on docker and then use it in github actions?

**Comment by vnvo2409** [comment]
I find this [tutorial](https://clubhouse.io/developer-how-to/how-to-set-up-a-hadoop-cluster-in-docker/). I think it should work. I will try it and will report to you later.

**Comment by kvignesh1420** [comment]
Okay.

**Comment by vnvo2409** [comment]
I would like to bring up the directory behavior of the cloud filesystem. To me, there's no use forcing `posix` behavior upon cloud filesystem, and it causes some misunderstoods for users ( https://github.com/tensorflow/tensorflow/issues/44884 ). If we can not have a change in this behavior, I think we should be more transparent with TF's users ( A warning or info that directory in cloud filesystem might not work as intented because the way that filesystem stores data is different + link )

**Comment by yongtang** [comment]
@vnvo2409 I think overall the `file systems` in tensorflow are not truly file systems that could completely replace standard file systems. Rather they are used to serve two purposes:
1. The file system (s3/gs) could be used in TF's C++ kernel for reading file content (e.g., `tf.io.read_file`)
2. The file system (s3/gs) could be used for saving and loading models. So in addition to reading file content, it also has the ability to get some file from a model directory. From that standpoint standard directory creation/deletion are not truly needed (only partially needed as long as the files under model directory could be created).
   In other words, I think a better documentation is needed but there is really no need to implement a feature-parity file system that could replace standard file system (or even replace gcs or s3's vendor provided python APIs in boto3 or google.cloud.).

**Comment by vnvo2409** [comment]
@yongtang Thank you for your reply ! I will also add it to my PR for the documentation.

**Comment by vnvo2409** [comment]
@yongtang 
Could you sum up what you are doing so our change doesn't overlap with each other ? Thank you.

**Comment by yongtang** [comment]
@vnvo2409 You can see the list of items that have not been done yet (https://github.com/tensorflow/io/issues/1183#issue-738561022):

Completed:
- [x] Initial S3 file system commit (PR #1191)
- [x] Initial GCS file system commit (PR #1203)
- [x] Initial HDFS file system commit (PR #1197)

Not started:
- [ ] [S3 file system] Port `s3_filesystem_test.cc` to python tests (https://github.com/tensorflow/io/pull/1191#issuecomment-726465162).
- [ ] [S3 file system] Bump AWS SDK to 1.8.x (https://github.com/tensorflow/io/pull/1191#issuecomment-726465162)
- [ ] [S3 file system] Fix multi download issue (https://github.com/tensorflow/io/pull/1191#issue-520111888)
- [ ] [S3 file system] Enable logging through TensorFlow C API (https://github.com/tensorflow/io/pull/1191#issue-520111888)
- [ ] [HDFS file system] Port `hadoop_filesystem_test.cc` to python tests
- [ ] [GCS file system] Port `gcs_filesystem_test.cc` to python tests

**Comment by vnvo2409** [comment]
@yongtang 
I could not see any `BUILD` file for the tests directory. Do we have to invoke the tests directly by `python` or am I missing something ?

**Comment by kvignesh1420** [comment]
@vnvo2409 we use `pytest` for testing the python based API's. For example:
```
TFIO_DATAPATH=bazel-bin python3 -m pytest -s -v tests/test_serialization_eager.py
```

**Comment by vnvo2409** [comment]
> [S3 file system] Enable logging through TensorFlow C API

I think this is already done via a0a4dc7ac7ecdfdc387a9709d5b923a4958c5096

**Comment by kvignesh1420** [comment]
@vnvo2409 I think so. Which work item are you planning to start with?

**Comment by vnvo2409** [comment]
I am working on `s3` test. I think it will be done by tomorrow.

**Comment by yongtang** [comment]
Thanks @vnvo2409 for the update 👍

**Comment by vnvo2409** [comment]
I found a filesystem named [gsmemcachedfs](https://github.com/tensorflow/io/tree/master/tensorflow_io/core/kernels/gsmemcachedfs). I think it has some what the same idea with the `ram_file_block_cache.h` in `gcs` plugins. In addition, it will stop working when we remove `gcs` filesystem out of `tensorflow` because it includes some headers from TensorFlow.
https://github.com/tensorflow/io/blob/0c9556ab981426a4a6668fee1ceedb2357241d49/tensorflow_io/core/kernels/gsmemcachedfs/memcached_file_system.h#L24

Should we do something with this filesystem ?

**Comment by yongtang** [comment]
@vnvo2409 Yes I think gsmemcachedfs will need to be ported to C modular file system API. /cc @marioecd

**Comment by vnvo2409** [comment]
@yongtang  @marioecd 
I think we should merge the library inside `gcs` plugins and `gsmemcachedfs` into one `ram_file_block_cache.h` and so on. gcs `ram_file_block_cache.h` is missing something like inheritance cause I am not aware that there is another filesystem depending on it. 

I could work on the migration but I might need something instructions for testing and general characters of that filesystem. WDYT ?

**Comment by jainakhil88** [comment]
Hi

Why is the issue not resolved. I am still facing same issue with latest tf-nightly-gpu build. I want to load pre-trained SSD model, I am trying to load model my trained model as explained from [here](https://github.com/nicknochnack/RealTimeObjectDetection/blob/main/Tutorial.ipynb).

Here is the complete error screen-shot
![image](https://user-images.githubusercontent.com/7304372/140751010-ddc7c1da-8a0b-4977-97c8-0b590ecba5cc.png)

_______________________________________________________________________
AlreadyExistsError                        Traceback (most recent call last)
<ipython-input-1-067428168487> in <module>
      2 from object_detection.utils import label_map_util
      3 from object_detection.utils import visualization_utils as viz_utils
----> 4 from object_detection.builders import model_builder

~\AppData\Roaming\Python\Python37\site-packages\object_detection\builders\model_builder.py in <module>
     35 from object_detection.meta_architectures import center_net_meta_arch
     36 from object_detection.meta_architectures import context_rcnn_meta_arch
---> 37 from object_detection.meta_architectures import deepmac_meta_arch
     38 from object_detection.meta_architectures import faster_rcnn_meta_arch
     39 from object_detection.meta_architectures import rfcn_meta_arch

~\AppData\Roaming\Python\Python37\site-packages\object_detection\meta_architectures\deepmac_meta_arch.py in <module>
     25 
     26 if tf_version.is_tf2():
---> 27   import tensorflow_io as tfio  # pylint:disable=g-import-not-at-top
     28 
     29 

C:\Anaconda3\envs\tfg2.4\lib\site-packages\tensorflow_io-0.21.0-py3.7-win-amd64.egg\tensorflow_io\__init__.py in <module>
     15 """tensorflow_io"""
     16 
---> 17 from tensorflow_io.python.api import *  # pylint: disable=wildcard-import
     18 from tensorflow_io.python.api.version import VERSION as __version__

C:\Anaconda3\envs\tfg2.4\lib\site-packages\tensorflow_io-0.21.0-py3.7-win-amd64.egg\tensorflow_io\python\api\__init__.py in <module>
     17 
     18 # tensorflow_io.core.python.ops is implicitly imported (along with file system)
---> 19 from tensorflow_io.python.ops.io_dataset import IODataset
     20 from tensorflow_io.python.ops.io_tensor import IOTensor
     21 

C:\Anaconda3\envs\tfg2.4\lib\site-packages\tensorflow_io-0.21.0-py3.7-win-amd64.egg\tensorflow_io\python\ops\__init__.py in <module>
     94 core_ops = LazyLoader("core_ops", "libtensorflow_io.so")
     95 try:
---> 96     plugin_ops = _load_library("libtensorflow_io_plugins.so", "fs")
     97 except NotImplementedError as e:
     98     warnings.warn("unable to load libtensorflow_io_plugins.so: {}".format(e))

C:\Anaconda3\envs\tfg2.4\lib\site-packages\tensorflow_io-0.21.0-py3.7-win-amd64.egg\tensorflow_io\python\ops\__init__.py in _load_library(filename, lib)
     62     for f in filenames:
     63         try:
---> 64             l = load_fn(f)
     65             if l is not None:
     66                 return l

C:\Anaconda3\envs\tfg2.4\lib\site-packages\tensorflow_io-0.21.0-py3.7-win-amd64.egg\tensorflow_io\python\ops\__init__.py in <lambda>(f)
     54         load_fn = lambda f: ctypes.CDLL(f, mode=ctypes.RTLD_GLOBAL)
     55     elif lib == "fs":
---> 56         load_fn = lambda f: tf.experimental.register_filesystem_plugin(f) is None
     57     else:
     58         load_fn = lambda f: tf.compat.v1.load_file_system_library(f) is None

C:\Anaconda3\envs\tfg2.4\lib\site-packages\tensorflow\python\framework\load_library.py in register_filesystem_plugin(plugin_location)
    176   """
    177   if os.path.exists(plugin_location):
--> 178     py_tf.TF_RegisterFilesystemPlugin(plugin_location)
    179 
    180   else:

AlreadyExistsError: File system for s3 already registered

_______________________________________________________________________

Below are my version details of all package in my environment [modules with versions.txt](https://github.com/tensorflow/io/files/7497335/modules.with.versions.txt) .

My machine details are
-Windows 10 
-
![image](https://user-images.githubusercontent.com/7304372/140751969-e63269a2-3636-425d-b4cd-d37826938e47.png)

Please let me know if you need anymore details to resolve the issue.

**Comment by vaibhavthapli** [comment]
> @mihaimaruseac One issue I noticed is that at the moment there is no way to `unload` the existing file systems embedded in TF. For example, if I am trying to register a shared object plugin for S3, the following error will thrown out because there is already a `s3` registered:
> 
> ```
> /Library/Python/3.8/site-packages/tensorflow/python/framework/load_library.py:178: in register_filesystem_plugin
>     py_tf.TF_RegisterFilesystemPlugin(plugin_location)
> E   tensorflow.python.framework.errors_impl.AlreadyExistsError: File system for s3 already registered
> ```
> 
> It looks like the error arises from the code in https://github.com/tensorflow/tensorflow/blob/dc3099c444d294b39cd79fe1d1a4bff59a0c6180/tensorflow/core/platform/env.cc#L82-L90
> 
> I am wondering if it makes sense to allow plugin registration to override existing static built file systems (s3/gcs/hdf5)? I think that will allow user to have more flexibilities. (We can lock `local`, `unix` file system to never be overridden).

I got the same error

**Comment by Sanjaygowda66** [comment]
> @mihaimaruseac One issue I noticed is that at the moment there is no way to `unload` the existing file systems embedded in TF. For example, if I am trying to register a shared object plugin for S3, the following error will thrown out because there is already a `s3` registered:
> 
> ```
> /Library/Python/3.8/site-packages/tensorflow/python/framework/load_library.py:178: in register_filesystem_plugin
>     py_tf.TF_RegisterFilesystemPlugin(plugin_location)
> E   tensorflow.python.framework.errors_impl.AlreadyExistsError: File system for s3 already registered
> ```
> 
> It looks like the error arises from the code in https://github.com/tensorflow/tensorflow/blob/dc3099c444d294b39cd79fe1d1a4bff59a0c6180/tensorflow/core/platform/env.cc#L82-L90
> 
> I am wondering if it makes sense to allow plugin registration to override existing static built file systems (s3/gcs/hdf5)? I think that will allow user to have more flexibilities. (We can lock `local`, `unix` file system to never be overridden).

Getting same error **tensorflow.python.framework.errors_impl.AlreadyExistsError: File system for s3 already registered** and **AlreadyExistsError: File system for az already registered**

**Comment by wansj** [comment]
> @mihaimaruseac One issue I noticed is that at the moment there is no way to `unload` the existing file systems embedded in TF. For example, if I am trying to register a shared object plugin for S3, the following error will thrown out because there is already a `s3` registered:
> 
> ```
> /Library/Python/3.8/site-packages/tensorflow/python/framework/load_library.py:178: in register_filesystem_plugin
>     py_tf.TF_RegisterFilesystemPlugin(plugin_location)
> E   tensorflow.python.framework.errors_impl.AlreadyExistsError: File system for s3 already registered
> ```
> 
> It looks like the error arises from the code in https://github.com/tensorflow/tensorflow/blob/dc3099c444d294b39cd79fe1d1a4bff59a0c6180/tensorflow/core/platform/env.cc#L82-L90
> 
> I am wondering if it makes sense to allow plugin registration to override existing static built file systems (s3/gcs/hdf5)? I think that will allow user to have more flexibilities. (We can lock `local`, `unix` file system to never be overridden).

I got this error when I tried to import the tensorflow_io module. The versions of tensorflow and  tensorflow_io are 2.6 and 0.20.0 respectively. Any solutions?
